Add security tests and fix some bugs in security implementation

dennis-adlink opened the pull request:
This PR adds a set of security tests and fixes a few bugs that were found when introducing these tests. 

The bugfixes in this PR:
- bb501ea The `decode_payload` function should return false in case encoded data is expected according to the security attributes and decoding fails
- 546a007 In `new_proxy_participant`, communication with a proxy participant should be set-up when there is any non-secured local participant. In case the remote participant allows unauthenticated participants in its security settings and advertises a non-secured topic, a reader of the local non-secure participant can read this data. This is fixed by adding the condition `!q_omg_participant_is_secure (pp)`. 
- b8a28dd Fix reader-writer synchronization in access control expiry_multiple test

New tests added (the test descriptions are also included as code comments):
- 93657cd / ad8dc89 Test discovery and liveliness protection by checking that `encode_datawriter_submessage` is called for `SEDP_BUILTIN_PUBLICATIONS_SECURE_WRITER` and `P2P_BUILTIN_PARTICIPANT_MESSAGE_SECURE_WRITER` depending on the discovery and liveliness protection settings in security configuration. 
- 269aaa8 Testing handshake result for any combination of protection kind values for rtps, discovery, liveliness, metadata (submsg) and payload protection settings. In all cases that have an encoding mismatch, the security handshake is expect to fail. 
- dc2071f Test read/write access control by running test cases with different combinations of allow and deny rules for publishing and subscribing a topic, and check correct working of the default policy.
- 4708b76 Test that all attributes and tokens retrieved from the access control plugin are correctly returned.
- 12f2e9a Check that communication for a topic that is allowed in the permissions config keeps working in case the publisher also creates a writer for a non-allowed topic
- d67436c Test communication for a non-secure participant with a secure participant that allows unauthenticated nodes in its governance configuration. Samples for a secured topic should not be received by a reader in the non-secure participant; samples for a non-secure topic should.
- 9080154 Validate that non-encrypted data will not be received by a reader that has protection enabled for rtps/submsg/payload. The test uses a crypto plugin wrapper mode that force the plugin to write plain data in the encoded output buffer to DDSI, ignoring the security attributes for the reader and writer.
- ba25f98 Check that all tokens that are sent to a remote participant are received correctly by that participant and the token-data stored in the remote participant is equal to the data in the token that was sent.

eboasson commented on src/security/core/tests/common/security_config_test_utils.c:
```diff
@@ -46,38 +58,43 @@ static const char *governance_xml =
+
+static const char *permissions_xml_allow_rule =
+    "      <allow_rule>"
+    "        <domains>${DOMAIN_ID:+<id>}${DOMAIN_ID:-<id_range><min>0</min><max>230</max></id_range>}${DOMAIN_ID:+</id>}</domains>"
```
Clever :)

eboasson commented on src/security/core/tests/common/cryptography_wrapper.c:
```diff
@@ -109,6 +125,219 @@ void set_entity_data_secret(struct dds_security_cryptography_impl * impl, const
+  }
+
+  struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->token_data_list), *elem = elem0;
+  while (elem != NULL)
```
`elem` can't be a null pointer here, I'd just a `do { ... } while (elem != elem0);`

eboasson commented on src/security/core/tests/common/cryptography_wrapper.c:
```diff
@@ -109,6 +125,219 @@ void set_entity_data_secret(struct dds_security_cryptography_impl * impl, const
+    return NULL;
+  }
+  struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->token_data_list), *elem = elem0;
+  while (elem != NULL)
```
See `get_crypto_tokens`

eboasson commented on src/security/core/tests/common/cryptography_wrapper.c:
```diff
@@ -109,6 +125,219 @@ void set_entity_data_secret(struct dds_security_cryptography_impl * impl, const
+  if (!ddsrt_circlist_isempty (&impl->encode_decode_log))
+  {
+    struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->encode_decode_log), *elem = elem0;
+    while (elem != NULL)
```
See `get_crypto_tokens`

eboasson commented on src/security/core/tests/common/cryptography_wrapper.c:
```diff
@@ -109,6 +125,219 @@ void set_entity_data_secret(struct dds_security_cryptography_impl * impl, const
+  if (!ddsrt_circlist_isempty (&impl->encode_decode_log))
+  {
+    struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->encode_decode_log), *elem = elem0;
+    while (elem != NULL)
```
See `get_crypto_tokens`

dennis-adlink commented on src/security/core/tests/common/access_control_wrapper.c:
```diff
@@ -63,8 +73,75 @@ struct dds_security_access_control_impl {
+  assert(impl->mode == PLUGIN_MODE_CHECK_RETURNS);
+  ddsrt_mutex_lock (&impl->returns_log_lock);
+  struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->returns_log), *elem = elem0;
+  while (elem != NULL)
```
I've fixed this, and the other loops over circular lists in crypto and access control plugins, in 7da139a.

dennis-adlink commented on src/security/core/tests/common/access_control_wrapper.c:
```diff
@@ -63,8 +73,75 @@ struct dds_security_access_control_impl {
+
+static void init_returns_log(struct dds_security_access_control_impl *impl)
+{
+  ddsrt_mutex_init (&impl->returns_log_lock);
```
`impl` is memset to 0 after allocation, but it's more clear to set this field to `false` explicitly, so I've added that in 7da139a.

dennis-adlink commented on src/security/core/tests/common/access_control_wrapper.c:
```diff
@@ -63,8 +73,75 @@ struct dds_security_access_control_impl {
+  ddsrt_mutex_unlock (&impl->returns_log_lock);
+}
+
+bool check_returns (struct dds_security_access_control_impl * impl)
```
You're right, it's confusing, I changed it in 7da139a.

dennis-adlink commented on src/security/core/tests/common/access_control_wrapper.c:
```diff
@@ -63,8 +73,75 @@ struct dds_security_access_control_impl {
 };
 
+
+static void init_returns_log(struct dds_security_access_control_impl *impl)
```
Done in 7da139a, and I've removed the specific test case for returns-checking as this is now checked in other tests that use the access control test wrapper in wrapper mode. This also revealed a bug in `is_topic_discovery_protected` where the return function call was missing for the topic attributes (which was not really an issue because the return implementation is currently a no-op).

dennis-adlink commented:
@eboasson Thanks for reviewing! I've force-pushed 221c8fd to remove the printfs that I added in the returns-logging in the access control plugin wrapper (these produced too much output and as a result it became hard to read other test output). I've also added 6ae0d30, which contains the fix (I think) for a race condition that could occur when permissions expire which causes the participant to be deleted, at the same the participant is deleted in the application (this race condition caused test failures in the access control test `permissions_expiry_multiple`)

eboasson commented:
Thanks @dennis-adlink, I agree with the changes. Indeed, before the introduction of permissions expiry there would never be multiple threads competing to delete the participant (because of how the API deals with entity deletion), but now that has changed.
